src/Paprika/Store/PagedDb.MultiHeadChain.cs

```diff
@@ -0,0 +1,977 @@
+                // potential optimization ahead
+                foreach (var (at, page) in batch.Changes)
+                {
+                    ref var slot =
+                        ref CollectionsMarshal.GetValueRefOrAddDefault(_pageTable, at, out var exists);
+
+                    if (exists == false)
+                    {
+                        // Does not exist, set and set the reverse mapping.
+                        slot = page;
+                        SetPageTableCache(at, page);
+
+                        _pageTableReversed[page] = at;
+                    }
+                    else if (ShouldOverwrite(page, slot))
+                    {
+                        // Remove previous reverse mapping
+                        _pageTableReversed.Remove(slot);
+
+                        // Override slot and set the reverse mapping
+                        slot = page;
+                        SetPageTableCache(at, page);
+                        _pageTableReversed[page] = at;
+                    }
+                }
+            }
+        }
+
+        public void Commit(uint blockNumber, in Keccak blockHash)
+        {
+            if (_chain == null)
+            {
+                throw new Exception(
+                    $"This head cannot be committed. " +
+                    $"It was created based on the empty root {Keccak.EmptyTreeHash} " +
+                    $"and should not be committed.");
+            }
+
+            MemoizeAbandoned();
+
+            // Copy the state hash
+            ParentHash = Root.Data.Metadata.StateHash;
+
+            SetMetadata(blockNumber, blockHash);
+
+            // The root ownership is now moved to the proposed batch.
+            // The batch is automatically leased by this head. It will be leased by the chain as well.
+            var batch = new ProposedBatch(_cowed.ToArray(), Root, ParentHash, Db._pool);
+
+            _cowed.Clear();
+            Clear();
+
+            // Create new root before it's proposed.
+            _root = CreateNextRoot(Root, _pool);
+
+            // Register proposal
+            var (reusePagesOlderThan, lastCommittedBatchId, read) = _chain.Propose(_read, batch);
+
+            // Locally track this proposed
+            _proposed.Enqueue(batch);
+
+            // Remove pages to be removed
+            // TODO: potentially make parallel, by gathering first keys to be removed and only then remove
+            while (_proposed.TryPeek(out var first) && first.Root.Header.BatchId <= lastCommittedBatchId)
+            {
+                var removed = _proposed.Dequeue();
+
+                foreach (var (at, page) in removed.Changes)
+                {
+                    if (_pageTable.TryGetValue(at, out var actual) && page.Equals(actual))
+                    {
+                        _pageTable.Remove(at);
+                        _pageTableReversed.Remove(actual);
+
+                        ref var cached = ref GetPageTableCacheSlot(at);
```

Comment:
I find it easier to reason about as it's structured now, because in the current take, we treat `_pageTable` as a source of truth and a cache only as an optimization. The time spent in this part is small when comparing to, for example, the application time. Also, most likely all the pages that are tried to be found will be found, so not sure if worth to do it the other way. Maybe we could optimize and `.Remove` always and try to fix when removed incorrectly? Still, the measured overhead was terribly small.